drv/stm32h7-hash/src/lib.rs

```diff
@@ -0,0 +1,364 @@
+            // XXX do i need to check DINIS? || !is_dinis_set() {
+            while self.is_busy() {
+                // || !is_dinis_set() {
+                hl::sleep_for(1);
+            }
+        }
+
+        if self.idx > 0 {
+            unsafe {
+                // Only the last block can have a partial word at the end.
+                // NBLW is initialized to 0 (last word has 32 valid bits) and
+                // can stay at zero if that doesn't change.
+                self.reg.str.modify(|_, w| {
+                    w.nblw()
+                        .bits(((self.count % SIZEOF_U32) * BITS_PER_BYTE) as u8)
+                });
+            }
+            for data in &self.block[0..self.idx] {
+                // If we were writing word instead of block at a time,
+                // then a busy check might be needed here.
+                unsafe {
+                    self.reg.din.write(|w| w.datain().bits(*data));
+                }
+            }
+            self.idx = 0;
+        }
+    }
+
+    fn write_word(&mut self, word: u32, valid_bytes: usize) {
+        if self.idx >= self.block.len() {
+            self.write_block();
+        }
+        self.block[self.idx] = word;
+        self.idx += 1;
+        self.count += valid_bytes;
+    }
+
+    // Little-endian data is fed to the hasher.
+    // e.g. "abc" is represented as 0x00636261
+    // Only the last data processed by the hasher can be less than 4 bytes.
+    pub fn update(&mut self, data: &[u8]) -> Result<(), HashError> {
+        match self.state {
+            State::Uninitialized => {
+                return Err(HashError::NotInitialized);
+            }
+            State::Initialized => {
+                self.state = State::Processing;
+            }
+            State::Processing => {}
+            _ => {
+                return Err(HashError::InvalidState);
+            }
+        };
+
+        // Incoming data may not be aligned.
+        //
+        // From the STM32H7 reference:
+        //
+        //  "...message string “abc” with a bit string representation of
+        //  “01100001 01100010 01100011” is represented by a 32-bit word
+        //  0x00636261, and 8-bit words 0x61626300."
+
+        // Deal with the remainder bytes from last update if any.
+        let mut offset = 0;
+        if self.nvalid > 0 {
+            while self.nvalid < 32 {
+                if offset >= data.len() {
+                    break;
+                }
+                self.remainder |= (data[offset] as u32) << self.nvalid;
+                self.nvalid += 8;
+                offset += 1;
+            }
+            if self.nvalid == 32 {
+                self.write_word(self.remainder, SIZEOF_U32);
+                self.nvalid = 0;
+                self.remainder = 0;
+            }
+        }
+
+        // Hash all of the whole words available.
+        // The words might not be aligned.
+        // TODO: How do the cool kids translate 4 unaligned bytes into a u32?
```

Comment:
You can try using the hideous strategy for slices [here](https://doc.rust-lang.org/std/primitive.u32.html#method.from_le_bytes). But this code looks fine to me.